packages/dev/core/src/Materials/Node/Blocks/Fragment/heightToNormalBlock.ts

```diff
@@ -67,34 +90,92 @@ export class HeightToNormalBlock extends NodeMaterialBlock {
             mat3 TBN = mat3(tangent, biTangent, normal);
-            vec3 worlddX = dFdx(position * 100.0);
-            vec3 worlddY = dFdy(position * 100.0);
-            vec3 crossX = cross(normal, worlddX);
-            vec3 crossY = cross(normal, worlddY);
-            float d = abs(dot(crossY, worlddX));
-            vec3 inToNormal = vec3(((((height + dFdx(height)) - height) * crossY) + (((height + dFdy(height)) - height) * crossX)) * sign(d));
-            inToNormal.y *= -1.0;
-            vec3 result = normalize((d * normal) - inToNormal);
-            return TBN * result;
-        }`;
+            `;
+
+        const endCode = this.generateInWorldSpace
+            ? ""
+            : `
+            result = TBN * result;
+            result = result * vec3(0.5) + vec3(0.5);
+            `;
+
+        const heightToNormal = `
+            vec4 heightToNormal(in float height, in vec3 position, in vec3 tangent, in vec3 normal) {
+                ${startCode}
+                ${this.automaticNormalizationTangent ? "tangent = normalize(tangent);" : ""}
+                ${this.automaticNormalizationNormal ? "normal = normalize(normal);" : ""}
+                vec3 worlddX = dFdx(position);
+                vec3 worlddY = dFdy(position);
+                vec3 crossX = cross(normal, worlddX);
+                vec3 crossY = cross(normal, worlddY);
+                float d = abs(dot(crossY, worlddX));
+                vec3 inToNormal = vec3(((((height + dFdx(height)) - height) * crossY) + (((height + dFdy(height)) - height) * crossX)) * sign(d));
+                inToNormal.y *= -1.0;
+                vec3 result = normalize((d * normal) - inToNormal);
+                ${endCode}
+                return vec4(result, 0.);
+            }`;
 
         state._emitExtension("derivatives", "#extension GL_OES_standard_derivatives : enable");
         state._emitFunction("heightToNormal", heightToNormal, "// heightToNormal");
         state.compilationString +=
             this._declareOutput(output, state) +
-            ` = heightToNormal(${this.input.associatedVariableName}, ${this.position.associatedVariableName}, ${this.tangent.associatedVariableName}.xyz, ${this.normal.associatedVariableName});\r\n`;
+            ` = heightToNormal(${this.input.associatedVariableName}, ${this.worldPosition.associatedVariableName}, ${
+                this.worldTangent.isConnected ? this.worldTangent.associatedVariableName : "vec3(0.)"
+            }.xyz, ${this.worldNormal.associatedVariableName});\r\n`;
+
+        if (this.xyz.hasEndpoints) {
+            state.compilationString += this._declareOutput(this.xyz, state) + ` = ${this.output.associatedVariableName}.xyz;\r\n`;
+        }
 
         return this;
     }
+
+    protected _dumpPropertiesCode() {
+        let codeString = super._dumpPropertiesCode();
+        codeString += `${this._codeVariableName}.generateInWorldSpace = ${this.generateInWorldSpace};\r\n`;
+

```

Comment:
you need to add the two automaticxxx as well

Reply:
Ah yes thanks!